compiler/rustc_span/src/def_id.rs

```diff
@@ -309,12 +309,47 @@ rustc_data_structures::define_id_collections!(LocalDefIdMap, LocalDefIdSet, Loca
 
 impl<CTX: HashStableContext> HashStable<CTX> for DefId {
     fn hash_stable(&self, hcx: &mut CTX, hasher: &mut StableHasher) {
-        hcx.hash_def_id(*self, hasher)
+        hcx.def_path_hash(*self).hash_stable(hcx, hasher);
+    }
+}
+
+impl<CTX: HashStableContext> HashStable<CTX> for LocalDefId {
+    #[inline]
+    fn hash_stable(&self, hcx: &mut CTX, hasher: &mut StableHasher) {
+        hcx.def_path_hash(self.to_def_id()).hash_stable(hcx, hasher);
     }
 }
 
 impl<CTX: HashStableContext> HashStable<CTX> for CrateNum {
     fn hash_stable(&self, hcx: &mut CTX, hasher: &mut StableHasher) {
-        hcx.hash_crate_num(*self, hasher)
+        hcx.def_path_hash(DefId { krate: *self, index: CRATE_DEF_INDEX }).hash_stable(hcx, hasher);
```

Comment:
```suggestion
        self.to_stable_hash_key(hcx).hash_stable(hcx, hasher);
```
Looks like this impl and a few impls above are all instances of
```rust
impl<CTX: HashStableContext> HashStable<CTX> for T where T: ToStableHashKey<CTX> {
    #[inline]
    fn hash_stable(&self, hcx: &mut CTX, hasher: &mut StableHasher) {
        self.to_stable_hash_key(hcx).hash_stable(hcx, hasher);
    }
}
```
Does coherence allow such an impl?
(If not, then it's probably better to implement all of them through `self.to_stable_hash_key(hcx).hash_stable(hcx, hasher)` anyway.)

Reply:
Coherence won't allow it because it can't forbid an implementation of `HashStable<local type>` in downstream crate.